nano/node/lmdb_txn_tracker.cpp

```diff
@@ -0,0 +1,126 @@
+#ifndef _WIN32
+#if !BEFORE_GNU_SOURCE
+#undef _GNU_SOURCE
+#endif
+#endif
+
+std::chrono::milliseconds constexpr nano::mdb_txn_tracker::min_time_held_open_ouput;
+
+nano::mdb_txn_tracker::mdb_txn_tracker (nano::logger_mt & logger_a, bool is_logging_database_locking_a) :
+logger (logger_a),
+is_logging_database_locking (is_logging_database_locking_a)
+{
+}
+
+void nano::mdb_txn_tracker::serialize_json (boost::property_tree::ptree & json, std::chrono::milliseconds min_time)
+{
+	// Copying is cheap compared to generating the stack trace strings, so reduce time holding the mutex
+	std::vector<mdb_txn_stats> copy_stats;
+	{
+		std::lock_guard<std::mutex> guard (mutex);
+		copy_stats = stats;
+	}
+
+	// Get the time difference now as creating stacktraces (Debug/Windows for instance) can take a while so results won't be as accurate
+	std::vector<std::chrono::milliseconds> times_since_start;
+	times_since_start.reserve (copy_stats.size ());
+	// clang-format off
+	std::transform (copy_stats.cbegin (), copy_stats.cend (), std::back_inserter (times_since_start), [] (const auto & stat) {
+		return stat.timer.since_start ();
+	});
+	// clang-format on
+	assert (times_since_start.size () == copy_stats.size ());
+
+	for (size_t i = 0; i < times_since_start.size (); ++i)
+	{
+		auto const & stat = copy_stats[i];
+		auto time_held_open = times_since_start[i];
+
+		if (time_held_open >= min_time)
+		{
+			nano::jsonconfig mdb_lock_config;
+
+			mdb_lock_config.put ("thread", stat.thread_name);
+			mdb_lock_config.put ("time_held_open", time_held_open.count ());
+			mdb_lock_config.put ("write", stat.is_write ());
+
+			boost::property_tree::ptree stacktrace_config;
+			for (auto frame : *stat.stacktrace)
+			{
+				nano::jsonconfig frame_json;
+				frame_json.put ("name", frame.name ());
+				frame_json.put ("address", frame.address ());
+				frame_json.put ("source_file", frame.source_file ());
+				frame_json.put ("source_line", frame.source_line ());
+				stacktrace_config.push_back (std::make_pair ("", frame_json.get_tree ()));
+			}
+
+			nano::jsonconfig stack (stacktrace_config);
+			mdb_lock_config.put_child ("stacktrace", stack);
+			json.push_back (std::make_pair ("", mdb_lock_config.get_tree ()));
+		}
+	}
+}
+
+void nano::mdb_txn_tracker::output_finished (nano::mdb_txn_stats & mdb_txn_stats)
+{
+	// Only output them if transactions were held for longer than a certain period of time
+	if (is_logging_database_locking && mdb_txn_stats.timer.since_start () >= min_time_held_open_ouput)
+	{
+		assert (mdb_txn_stats.stacktrace);
+		logger.always_log (boost::str (boost::format ("%1%ms %2% held on thread %3%\n%4%") % mdb_txn_stats.timer.since_start ().count () % (mdb_txn_stats.is_write () ? "write lock" : "read") % mdb_txn_stats.thread_name % *mdb_txn_stats.stacktrace));
+	}
+}
+
+void nano::mdb_txn_tracker::add (const nano::transaction_impl * transaction_impl)
+{
+	std::lock_guard<std::mutex> guard (mutex);
+	stats.emplace_back (transaction_impl);
```

Comment:
This shows up in the profiler, taking ~50% of the time spent in starting a read tx. I suspect it's the instantiation of the stack trace object induced by the emplace. Maybe a flag to enable/disable the tracking feature is needed.

Reply:
Made a new config node "diagnostics" which enables turning this on/off. Have made read/write minimum time open for tracking options configurable using this as well (as opposed to be just a hardcoded single value for both). While this may appear more suited to a CLI debug command as it is most useful to developers, I think a config node option makes more sense as we are likely going to want this enabled for all of our runs (rather than having to remember to set it every single time) as we can't retroactively find out what went wrong afterwards when there is ledger bloat. I imagine we will find more uses for this config in the future as well.